### PR TITLE
Jetpack Plans: Add to redux store & actions for the automated plugin install flow

### PR DESCRIPTION
**Pull request by ryelle**
This adds a new item to the redux store, `premium`, which keeps track of premium services for jetpack sites. This is used in the automated install flow to step through the installation/activation/configuration of each plugin in a user-facing checklist.

`fetchInstallInstructions` gets the list of plugins to install from wp.com API, as this can differ based on plan (and also allows for future flexibilty).

Once we have the plugin list, we can start installing them. The install action also triggers the rest of the status actions (activate etc), to make this as automated as possible. Since the main plugins work is still in flux stores, we need to work with the flux dispatcher as well as redux's dispatch.

This PR includes tests for the selectors and reducer.

_(currently not really configuring the plugin after activation, but that can be a separate PR)_

cc @johnHackworth @roccotripaldi

### CODE REVIEW

File: client/state/plugins/premium/actions.js [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+		activate( site, data, dispatch );
+	} );
+};
+
+function update( site, plugin, dispatch ) {
+	wpcom.pluginsUpdate( site.ID, plugin.id, ( error ) => {
+		if ( error ) {
+			console.log( plugin.slug );
+			console.log( 'Update Failed', error.name );
```
Comment (johnHackworth) [inline client/state/plugins/premium/actions.js]: piiing!

File: client/state/plugins/premium/actions.js [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+		} );
+
+		activate( site, plugin, dispatch );
+	} );
+};
+
+function activate( site, plugin, dispatch ) {
+	wpcom.pluginsActivate( site.ID, plugin.id, ( error, data ) => {
+		if ( error && error.name !== 'ActivationErrorError' ) {
+			console.log( 'Activate Failed:', plugin.id, error.name );
```
Comment (johnHackworth) [inline client/state/plugins/premium/actions.js]: See previous ping :D

File: client/state/plugins/premium/reducer.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+		case PLUGIN_SETUP_FETCH_INSTRUCTIONS:
+			return Object.assign( {}, state, { [ action.siteId ]: true } );
+		case PLUGIN_SETUP_RECEIVE_INSTRUCTIONS:
+			return Object.assign( {}, state, { [ action.siteId ]: false } );
+		default:
+			return state;
+	}
+}
+
+function pluginStatus( state, action ) {
```
Comment (johnHackworth) [inline client/state/plugins/premium/reducer.js]: Can we add some individual comments explaining what each one of this functions do? I'm struggling to understand how it is divided and what's each one do

Comment (ryelle) [inline client/state/plugins/premium/reducer.js]: Yeah there's a lot of nested functions here - I've pushed an update with some comments & a human-readable re-ordering of functions. If you haven't already seen this, I have an example state in the [readme](https://github.com/Automattic/wp-calypso/blob/add/premium-service-redux/client/state/plugins/premium/README.md#reducer) which might help.

File: client/state/plugins/premium/test/reducer.js [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+	PLUGIN_SETUP_INSTALL,
+	PLUGIN_SETUP_ACTIVATE,
+	PLUGIN_SETUP_CONFIGURE,
+	PLUGIN_SETUP_FINISH,
+	PLUGIN_SETUP_ERROR
+} from 'state/action-types';
+
+import { isRequesting, plugins } from '../reducer';
+
+const initSite = [
```
Comment (johnHackworth) [inline client/state/plugins/premium/test/reducer.js]: I think it would be better if all this constants can be extracted to a helper file

File: client/state/action-types.js [hunk 5]
```diff
@@ -67,6 +67,15 @@ export const PAGE_LINK_ADD = 'PAGE_LINK_ADD';
 export const PAGE_META_ADD = 'PAGE_META_ADD';
 export const PAGE_TITLE_SET = 'PAGE_TITLE_SET';
 export const PAGE_UNREAD_COUNT_SET = 'PAGE_UNREAD_COUNT_SET';
+export const PLUGIN_SETUP_FETCH_INSTRUCTIONS = 'PLUGIN_SETUP_FETCH_INSTRUCTIONS';
```
Comment (gwwar) [inline client/state/action-types.js]: It reads a bit funny, but our naming convention suffixes the verb. So this might be `PLUGIN_SETUP_INSTRUCTIONS_FETCH` https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#actions

File: client/state/plugins/premium/reducer.js [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+ * Tracks the requesting state for premium plugin "instructions" (the list
+ * of plugins and API keys) on a per-site index.
+ */
+export function isRequesting( state = false, action ) {
+	switch ( action.type ) {
+		case PLUGIN_SETUP_FETCH_INSTRUCTIONS:
+			return Object.assign( {}, state, { [ action.siteId ]: true } );
+		case PLUGIN_SETUP_RECEIVE_INSTRUCTIONS:
+			return Object.assign( {}, state, { [ action.siteId ]: false } );
+		default:
```
Comment (gwwar) [inline client/state/plugins/premium/reducer.js]: I should add a warning for this, but each reducer should also implement `SERIALIZE` and `DESERIALIZE` (see [data persistence](https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#data-persistence--2754-) ) Without these cases, this subtree will be persisted by default. If you're not ready to persist state here, return the default state, which in this case is `false`.

File: client/state/plugins/premium/actions.js [hunk 7]
```diff
@@ -0,0 +1,251 @@
+/**
+* External dependencies
+*/
+let wpcom = require( 'lib/wp' ).undocumented();
```
Comment (gwwar) [inline client/state/plugins/premium/actions.js]: This can be `const`. :)

File: client/state/plugins/premium/actions.js [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+};
+
+function configure( site, plugin, dispatch ) {
+	setTimeout( () => {
+		dispatch( {
+			type: PLUGIN_SETUP_FINISH,
+			siteId: site.ID,
+			slug: plugin.slug,
+		} );
+	}, 1500 );
```
Comment (gwwar) [inline client/state/plugins/premium/actions.js]: Why do we need such a long timeout?

File: client/tests.json [hunk 9]
```diff
@@ -495,6 +495,9 @@
 			"test": [ "actions", "reducer" ]
 		},
 		"plugins": {
+			"premium": {
+				"test": [ "reducer", "selectors" ]
```
Comment (gwwar) [inline client/tests.json]: No action tests?

File: client/state/plugins/premium/test/reducer.js [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+import { initSite, installingSite, activatingSite, configuringSite, finishedPluginSite, siteWithError } from './examples';
+
+describe( 'premium reducer', () => {
+	describe( 'isRequesting', () => {
+		it( 'should track when fetches start', () => {
+			const state = isRequesting( undefined, {
+				type: PLUGIN_SETUP_FETCH_INSTRUCTIONS,
+				siteId: 'one.site'
+			} );
+			expect( state ).to.deep.equal( { 'one.site': true } );
```
Comment (gwwar) [inline client/state/plugins/premium/test/reducer.js]: To avoid the typing next time,  you can use `.eql` for deep equals in mocha.

File: client/state/plugins/premium/selectors.js [hunk 11]
```diff
@@ -0,0 +1,54 @@
+import find from 'lodash/find';
+import filter from 'lodash/filter';
+
+const isRequesting = function( state, siteId ) {
+	// if the `isRequesting` attribute doesn't exist yet,
+	// we assume we are still launching the fetch action, so it's true
+	if ( typeof state[ siteId ] === 'undefined' ) {
+		return true;
+	}
+	return state[ siteId ];
```
Comment (gwwar) [inline client/state/plugins/premium/selectors.js]: Our current convention is to pass through the full state object to our selectors. We need to tidy up the interfaces of the other existing plugin selectors, in a future PR. See also [selectors](https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#selectors)

So this might look like:
```javascript
return state.plugins.premium[ stateId ];
```

File: client/state/plugins/premium/test/reducer.js (hunk 4 above)
Comment (gwwar) [inline client/state/plugins/premium/test/reducer.js]: ;) I personally like these inline since it reduces context switching. It's fine in this case. If these test fixtures get too large we can maybe consider adding in some helper test factories.

File: client/state/plugins/premium/actions.js (hunk 8 above)
Comment (ryelle) [inline client/state/plugins/premium/actions.js]: Since it's an action on the user's site it could take some time. That said, this will be going away in another PR shortly after this gets merged (when the real configuration step is implemented).

File: client/state/plugins/premium/actions.js [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+			site: site,
+			plugin: plugin,
+			data: null,
+			error: error
+		} );
+	} );
+}
+
+function autoupdate( site, plugin ) {
+	getPluginHandler( site, plugin.id ).enableAutoupdate();
```
Comment (johnHackworth) [inline client/state/plugins/premium/actions.js]: shouldn't we catch any errors here too? Even if it's just for showing a warning to the user?

File: client/state/plugins/premium/actions.js [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+			break;
+		case 'akismet':
+			option = 'wordpress_api_key';
+			break;
+		case 'polldaddy':
+			option = 'polldaddy_api_key';
+			break;
+	}
+	if ( ! option || ! plugin.key ) {
+		let optionError = new Error( 'We can\'t configure this plugin.' );
```
Comment (johnHackworth) [inline client/state/plugins/premium/actions.js]: is this error message going to be shown to the user? I can't find any place where it's being printed out, but maybe we plan to it sooner or later? In that case, we should translate it

**Comment by johnHackworth** [comment]
I left a couple of minor comments, but they can be addressed (if needed) in a new PR . I think this is ready.I've rebased it with master already, so feel free to merge it :shipit:

File: client/state/plugins/premium/actions.js (hunk 12 above)
Comment (ryelle) [inline client/state/plugins/premium/actions.js]: @johnHackworth for the plugins auto-config flow, not really - showing the user an error on this step would just be confusing, i think. ideally, when plugins goes redux, these functions will end up going away for the more generic plugins install actions, and those would handle any errors.
